docs/designs/checkpoint_design.md

```diff
@@ -0,0 +1,59 @@
+# ElasticDL Training Checkpoints Design
+This document describes the design of training checkpoints for ElasticDL
+
+## Motivation
+Checkpoints capture the exact value of all parameters (tf.Variable objects and embedding vectors) used by a model. In ElasticDL, we need to save all parameters of the model to checkpoints during training for the following reasons:
+* We can export the model with the best evaluation metrics during training by checkpoints. Generally, we evaluate the model and monitor the evaluation metrics during training. We will stop training when we find the evaluation metrics don't go better or even go worse due to over-fitting. So, the final parameters may not be the best. With checkpoints, we can choose a checkpoint with the best evaluation metrics to export for serving.
+* We can restore the model parameters from a checkpoint to train or predict. Fault tolerance in ElasticDL can not resolve all unexpected error e.g. a power outage and OS fault in the master. We are going to have a bad time if we lose one or more of our experiments due to those errors. Other times, even if we don't experience an unforeseen error, we might just want just to resume a particular state of the training for a new experiment or try different things from a given state.
+
+In the following sections, we will describe the design of how to make a checkpoint and restore model parameters from a checkpoint in ElasticDL.
+
+## Design Components
+There are two design components: export and restore.
+
+### Export Model Parameters to a Checkpoint
+
+**When to save a checkpoint during training.** 
+
+In ElasticDL, we will save all model parameters to a checkpoint directory every `checkpoint_steps`. Though, some frameworks support to save checkpoints every  steps or every seconds, e.g. `tf.estimator`. Using ParameterServerStrategy in ElasticDL, PS instances are asynchronous and we cannot guarantee that the iteration steps of all PS instances are consistent at the same time. So, we don't support to save checkpoints every seconds.
+
+
+**Where to save a checkpoint.** 
+
+We need to set the `checkpoint_dir` in args if we want to save checkpoints. The saved checkpoint directory will be generated by `checkpoint_dir` and iteration steps which we call model version in ElasticDL, e.g. `/{checkpoint_dir}/model_v{version}/`.
+```python
+checkpoint_folder_name = "model_v%s" % version
+checkpoint_dir = os.path.join(
+    args.checkpoint_dir, checkpoint_folder_name
+)
+```
+ElasticDL is a Kubernetes-native framework and local directory in the pod will be lost if the pod exits in Kubernetes. So the `checkpoint_dir` should be a persistent directory which will not be lost after the pod exit. Now, we can utilize `hostPath` and `PersistentVolume` to get a persistent directory to save checkpoints in ElasticDL.
```

Comment:
It seems that we should save the checkpoint to shared storage. Are`hostPath` and `PersistentVolumn` used to mount a shared storage?

Reply:
Yes, `hostPath` can be used to mount the local directory on the node in minikube. And `PersistentVolumn` can be used to mount glusterfs in Kubernetes cluster.

Reply:
How about add the link of hostPath and PersistentVolume in this doc?

Reply:
Yeah, good suggestion!